lib/Providers/OpenIDConnectProvider.php

```diff
@@ -0,0 +1,186 @@
+        }
+    }
+
+    protected function createResourceOwner(array $response, AccessToken $token)
+    {
+        return new GenericResourceOwner($response, 'id');
+    }
+
+    /**
+     * Do any required verification of the id token and return an array of decoded claims
+     *
+     * @param string $id_token Raw id token as string
+     * @return array associative array of claims decoded from the id token
+     */
+    public function verifyIdToken($id_token)
+    {
+        try {
+            $keys = $this->getSigningKeys();
+            $claims = JWT\JWT::decode($id_token, $keys, ['RS256']);
+            if ($claims->aud !== $this->clientId) {
+                throw new IdentityProviderException("ID token has incorrect audience", 0, $claims->aud);
+            }
+            if ($claims->iss !== $this->issuer) {
+                throw new IdentityProviderException("ID token has incorrect issuer", 0, $claims->iss);
+            }
+        } catch (\UnexpectedValueException $e) {
+            throw new IdentityProviderException("ID token validation failed", 0, $e->getMessage());
+        }
+    }
+
+    protected function prepareAccessTokenResponse(array $result)
+    {
+        $result = parent::prepareAccessTokenResponse($result);
+        $this->verifyIdToken($result['id_token']);
+        return $result;
+    }
+
+    protected function getOpenIDConfiguration() {
+        if (isset($this->openIdConfiguration)) {
+            return $this->openIdConfiguration;
+        }
+
+        $config = $this->getParsedResponse($this->getRequest('GET', $this->issuer . self::CONFIGURATION_PATH));
+        $this->openIdConfiguration = $config;
+        return $config;
+    }
+
+    public static function base64url_decode($input) {
+        return base64_decode(strtr($input, '-_', '+/'));
+    }
+
+    protected function getSigningKeys() {
+        $url = $this->getOpenIDConfiguration()['jwks_uri'];
+        $jwks = $this->getParsedResponse($this->getRequest('GET', $url));
+        $keys = [];
+        foreach ($jwks['keys'] as $key) {
+            $kid = $key['kid'];
+            if (array_key_exists('x5c', $key)) {
+                $x5c = $key['x5c'];
+                $keys[$kid] = "-----BEGIN CERTIFICATE-----\n" . $x5c[0] . "\n-----END CERTIFICATE-----";
+            } else if ($key['kty'] === 'RSA') {
+                $e = self::base64url_decode($key['e']);
+                $n = self::base64url_decode($key['n']);
+                $keys[$kid] = \RobRichards\XMLSecLibs\XMLSecurityKey::convertRSA($n, $e);
+            } else {
+                error_log("Failed to load key data for key id: " . $kid);
+            }
+        }
+        return $keys;
+    }
+
+    /**
+     * Returns the base URL for authorizing a client.
+     *
+     * Eg. https://oauth.service.com/authorize
+     *
+     * @return string
+     */
+    public function getBaseAuthorizationUrl()
+    {
+        return $this->getOpenIDConfiguration()["authorization_endpoint"];
```

Comment:
It feels a little bit weird that we need to go through `getOpenIDConfiguration()` every time we need to do something with the configuration. Would it be an option to initialize the `$openIdConfiguration` property in the constructor, and then refer always to `$this->openIdConfiguration` instead? Are there any other entrypoints to the class apart from the constructor?

Reply:
`getOpenIDConfiguration()` does a http fetch. I personally does not like to do IO in constructors. It makes for ugly code, and also I think it might be possible to use this class without requiring that configuration in some cases and then it would be a waste of resources to do the network IO in the constructor.

Reply:
That's a good point. If it's possible to use this class in some way without ever loading a OIDC configuration from the OP, then it makes sense to keep it like it is 👍